### DISCUSSION TITLE
TypeError: can't create 'module' instances

### DISCUSSION THREAD

**Discussion by klaasbrant** [micropython/micropython#9757] [Libraries & Drivers] [answered]
I did write some micropython classes to set the RTC with a NTP server and a class to handle the timezone time including handling of daylight saving time. Then I thought why not base the last one on the buildin time module class so localtime finally get a correct value. But that doesn't work:

import utime
class EZTime(utime):
    def __init__(self,tz=""):
        self.timezone(tz)
        super.__init__()

Traceback (most recent call last):
  File "<stdin>", line 3, in <module>
TypeError: can't create 'module' instances

Is this a micropython restriction? Micropython classes cannot extend a class in a C language module?

**Comment by stinos** [comment]
> Micropython classes cannot extend a class in a C language module?

That is possible, but you're trying to inherit a module, not a class. There's no such thing as a 'module class' (as far as I'm aware). I'm not sure even CPython can do that. You can create a new module which hass all members of another module (e.g. https://softwareengineering.stackexchange.com/questions/298019/how-to-achieve-inheritance-when-using-just-modules-and-vanilla-functions-in-pyth). Or you can inherit a class. Btw isn't `timezone` in the `datetime` module?

**Comment by klaasbrant** [comment]
There is a datetime in RTC (ESP32) but no timezone in there. Just a simple class to obtain a datetime tuple.

The stack overflow  topic discusses function inside a module. But utime is a class not a function.
You can do dir(utime) and it shows the fact that it is a class and all it's methods.
The stack overflow is right a from <xyz> import * gives access to all the functions, classes and variables in <xyz>
And if you do that and then create your own function with the same name it will overwrite the imported one.
That is why import * is quick and dirty and might backfire one day.

doing:

from utime import *
class EZTime(utime):
    def __init__(self,tz=""):
        self.timezone(tz)
        super.__init__()

gives the same error. So it is not an import problem... As if you cannot extend a class from the c-language module in micropython?

**Reply by jimmo** [reply]
@stinos is right, you cannot do this in CPython either:

```
>>> import time
>>> class X(time):
...   pass
... 
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
TypeError: module() takes at most 2 arguments (3 given)
```

**Comment by stinos** [comment, marked as answer]
> You can do dir(utime) and it shows the fact that it is a class and all it's methods.

`dir` just shows members, that alone doesn't prove it's a class. I don't know what platform you're on, but don't you get

```
>>> import utime
>>> utime
<module 'utime'>
```

?

As opposed to for example
```
>>> list
<class 'list'>
```

**Reply by klaasbrant** [reply]
OMG so utime is just a module with a bunch of functions?! Confusing, some are classes like RTC, I2C, PWM and some are not like utime.  I will create methods with similar names for my function because I hate that you depend on the import order. 

I thought utime was a class because dir showed __class__ I wonder where that comes from?  Next time I will do a type(<name>) before I ask questions ;-)

**Reply by stinos** [reply]
>  thought utime was a class because dir showed class I wonder where that comes from?

Ah, `__class__`, yes I can see how that could cause confusion. That's essentially the same functionality as `type(<something>)`.
